### PR TITLE
Performances improvements of ROCAUC

### PR DESCRIPTION
**Pull request by vfdev-5**
### Description
Performances improvements of ROCAUC
- detach predictions to break graph building
- clone and store predictions and targets on cpu

Related to https://github.com/pytorch/ignite/pull/967

### Status
Ready

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### CODE REVIEW

File: monai/handlers/roc_auc.py [hunk 1]
```diff
@@ -65,12 +65,14 @@ def update(self, output: Sequence[torch.Tensor]):
         if y.ndimension() not in (1, 2):
             raise ValueError("targets should be of shape (batch_size, n_classes) or (batch_size, ).")
 
-        y_pred = y_pred.to(self._predictions)
-        y = y.to(self._targets)
+        y_pred = y_pred.detach().to(dtype=torch.float32, device="cpu").clone()
+        y = y.to(dtype=torch.long, device="cpu").clone()
```
Comment (Nic-Ma) [inline monai/handlers/roc_auc.py:69]: 1. I think we don't need to `detach` because the typical usage of metrics is in the validation with auto_grad=False.
If you faced problems with `detach` about metrics before, welcome to share and discuss.
2. No need to move to CPU as `compute_roc_auc ` can support GPU tensors and slightly faster.

Comment (vfdev-5) [reply monai/handlers/roc_auc.py:69]: @Nic-Ma thanks for the review! I would like to discuss a bit more these points:

Concerning the point 1, yes, you are right that with `torch.no_grad()` which decorates `iteration_completed` and clone, there is no grads any more.

Concerning the point 2, yes, that's true that on GPU it will be faster. However, maybe we also need to keep in mind that storage on GPU can bother users with large models and large validation datasets. Well, this is still rather relative consumption. As predictions and targets are tensors of ndim 1 or 2, so at most we can imagine to store 2 tensors of size (N, K), e.g. N=1e6, K=10 and this will cost ~100-200 MB on GPU.
Another point is worth also to keep in mind is about DDP (in ignite we are not yet done that), to make it work with DDP, we will need to gather all predictions and targets either on update or on compute.

What do you think ? Anyway, I'll remove `detach`.

Comment (Nic-Ma) [reply monai/handlers/roc_auc.py:69]: 1. Classification task result is not very big, for example 200MB as your said, it's not a problem to put on CPU or GPU, and CPU memory is also fully occupied when doing 3D medical image transforms. So let's keep it in the original `y_pred`?
2. DDP is a good point, thanks for your reminder! When we add support for DDP, I will need to totally check all things, not only AUC, thanks.
3. Do we really need to `clone()` the `y_pred` and `y`? We don't modify the values in AUC. Do you have any example that `y_pred` is modified somewhere after storing to the prediction list of AUC and so affect AUC computation?

Thanks.

Comment (vfdev-5) [reply monai/handlers/roc_auc.py:69]: Sounds good for "1.", I'll remove `to(device='cpu')`

About "3.", well, it will be safer to clone. As user may want to compute metrics on `y_pred` and then perform some inplace ops... Also, if we remove `clone`, we need to put `detach`, otherwise graph is not stopped.

Comment (Nic-Ma) [reply monai/handlers/roc_auc.py]: 1. OK, you are right. `clone` is safer as this AUC handler stores data and compute when epoch completed, not like other metrics that compute immediately.
2. Thanks for your update about `to(cpu)`, seems `to(dtype=XXX)` is unnecessary now as `compute_roc_auc` can support different data types.
3. Others look good to me now.

Comment (vfdev-5) [reply monai/handlers/roc_auc.py]: About "2.", I can remove it. However, we need to keep in mind that `torch.cat` can fail if user somehow mixes the types of either `y_pred` or `y`. And the error will appear only in the end of validation phase.

Comment (Nic-Ma) [reply monai/handlers/roc_auc.py]: Thanks for your reminder, let's keep in mind that and remove the `to(dtype)`. I believe in most cases, the output data of every iteration is in the same data type and format.
Unnecessary `to(type)` will potentially copy data.
Do you agree?
Thanks.

Comment (vfdev-5) [reply monai/handlers/roc_auc.py]: > Unnecessary to(type) will potentially copy data.

If types are the same, there is no copy, but if they are not the same there is one.
